src/unix/bsd/freebsdlike/freebsd/mod.rs

```diff
@@ -1396,28 +1437,62 @@ cfg_if! {
+                    && match self.sigev_notify {
+                        ::SIGEV_NONE => true,
+                        ::SIGEV_SIGNAL => true,
+                        ::SIGEV_THREAD => unsafe {
+                            self._sigev_un._sigev_thread
+                                == other._sigev_un._sigev_thread
+                        },
+                        ::SIGEV_KEVENT => unsafe {
+                            self._sigev_un._kevent_flags
+                                == other._sigev_un._kevent_flags
+                        },
+                        ::SIGEV_THREAD_ID => unsafe {
+                            self._sigev_un._threadid
+                                == other._sigev_un._threadid
+                        },
+                        _ => false
+                    }
             }
         }
         impl Eq for sigevent {}
         impl ::fmt::Debug for sigevent {
             fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
-                f.debug_struct("sigevent")
-                    .field("sigev_notify", &self.sigev_notify)
-                    .field("sigev_signo", &self.sigev_signo)
-                    .field("sigev_value", &self.sigev_value)
-                    .field("sigev_notify_thread_id",
-                           &self.sigev_notify_thread_id)
-                    .finish()
+                let mut ds = f.debug_struct("sigevent");
+                ds.field("sigev_notify", &self.sigev_notify);
+                ds.field("sigev_signo", &self.sigev_signo);
+                ds.field("sigev_value", &self.sigev_value);
+                // The sigev_notify field indicates which union fields are valid
+                unsafe {
+                    match self.sigev_notify {
+                        ::SIGEV_THREAD => ds.field("_sigev_thread",
+                            &self._sigev_un._sigev_thread),
+                        ::SIGEV_KEVENT => ds.field("_kevent_flags",
+                            &self._sigev_un._kevent_flags),
+                        ::SIGEV_THREAD_ID => ds.field("_threadid",
+                            &self._sigev_un._threadid),
+                        _ => &mut ds
+                    }
+                };
+                ds.finish()
             }
         }
         impl ::hash::Hash for sigevent {
             fn hash<H: ::hash::Hasher>(&self, state: &mut H) {
                 self.sigev_notify.hash(state);
                 self.sigev_signo.hash(state);
                 self.sigev_value.hash(state);
-                self.sigev_notify_thread_id.hash(state);
+                // The sigev_notify field indicates which union fields are valid
+                unsafe {
+                    match self.sigev_notify {
+                        ::SIGEV_THREAD => self._sigev_un._sigev_thread.hash(state),
+                        ::SIGEV_KEVENT => self._sigev_un._kevent_flags.hash(state),
+                        ::SIGEV_THREAD_ID => self._sigev_un._threadid.hash(state),
```

Comment:
Same here.

I think either these impls need to be removed because `sigev_notify` is not trustworthy, or constructing `sigevent` with a user-controlled `sigev_notify` needs to be unsafe in some way, or constructing `sigevent` needs to be disallowed altogether.